src/basic/chattr-util.c

```diff
@@ -172,3 +173,79 @@ int read_attr_at(int dir_fd, const char *path, unsigned *ret) {
 
         return read_attr_fd(fd, ret);
 }
+
+int get_proj_id(int fd, uint32_t *ret) {
+        struct fsxattr attrs;
+        _cleanup_close_ int fd_reopened = -EBADF;
+
+        assert(ret);
+
+        fd = fd_reopen_condition(fd, 0, O_PATH, &fd_reopened);
+        if (fd < 0)
+                return fd;
+
+        if (ioctl(fd, FS_IOC_FSGETXATTR, &attrs) < 0)
+                return -errno;
+
+        *ret = attrs.fsx_projid;
+
+        return 0;
+}
+
+int set_proj_id(int fd, uint32_t proj_id) {
+        struct fsxattr attrs;
+        _cleanup_close_ int fd_reopened = -EBADF;
+
+        fd = fd_reopen_condition(fd, 0, O_PATH, &fd_reopened);
+        if (fd < 0)
+                return fd;
+
+        if (ioctl(fd, FS_IOC_FSGETXATTR, &attrs) < 0)
+                return -errno;
+
+        struct stat statbuf;
+        if (fstat(fd, &statbuf) < 0)
+                return -errno;
+
+        if (attrs.fsx_projid == proj_id) {
+                if (!S_ISDIR(statbuf.st_mode))
+                        return 0;
+                else if (attrs.fsx_xflags == FS_XFLAG_PROJINHERIT)
+                        return 0;
+        }
```

Comment:
merge into:

```c
if (atrs.fsx_projid == proj_id && (!S_ISDIR(statbuf.st_mode) || attrs.fsx_xflags == FS_XLFAG_PROJINHERIT))
       return 0;
``

Reply:
or, uh, checking for the flag via `==` is broken? should be via & (or FLAG_SET()...